src/fileHandler.ts

```diff
@@ -0,0 +1,76 @@
+import * as vscode from 'vscode';
+import * as path from 'path';
+import * as fs from 'fs';
+import * as namespace from './namespace';
+
+function isCljFile(uri: vscode.Uri): boolean {
+  return uri.fsPath && ((uri.fsPath.endsWith(".clj")
+    || uri.fsPath.endsWith(".cljs")
+    || uri.fsPath.endsWith(".cljc")));
+}
+
+function isPrefix(parentPath: string, filePath: string): boolean {
+  const relative = path.relative(parentPath, filePath);
+  return !relative || (!relative.startsWith('..') && !path.isAbsolute(relative));
+}
+
+function resolveNsName(sourcePaths: string[], filePath: string): string {
+  if (!sourcePaths || !sourcePaths.length) {
+      return "";
+  }
+  for (const sourcePath of sourcePaths) {
+      if (isPrefix(sourcePath, filePath)) {
+          const relative = path.relative(sourcePath, path.dirname(filePath));
+          return relative.replace(/[\/\\]/g, ".")
+                         .replace("_", "-"); // Is such style acceptable?
+      }
+  }
+  return "";
+}
+
+function resolveFileName(filePath: string): string {
+  const fileName: string = path.basename(filePath);
+  const extName: string = path.extname(fileName);
+  return fileName.substring(0, fileName.length - extName.length)
+                 .replace("_", "-");
+}
+
+export default async function handleNewCljFiles(e: vscode.FileCreateEvent) {
+  const emptyFiles: vscode.Uri[] = [];
+  const textDocuments: vscode.TextDocument[] = [];
+  for (let file of e.files) {
+      if (!isCljFile(file)) {
+          continue;
+      }
+      const textDocument = await vscode.workspace.openTextDocument(file);
+      if (textDocument.getText()) { // Ignore the non-empty files.
+          continue;
+      }
+      emptyFiles.push(file);
+      textDocuments.push(textDocument);
+  }
+  if (!emptyFiles.length) {
+      return;
+  }
+  let sourcePaths: string[] = [];
+  const client = namespace.getSession();
+  if (client) {
+      const resp = await client.classpath();
+      if (resp && resp.classpath) {
+        sourcePaths = resp.classpath.filter(p => fs.existsSync(p) && fs.lstatSync(p).isDirectory());
+      }
+  }
+  for (let i = 0; i < emptyFiles.length; i++) {
+      const filename: string = resolveFileName(textDocuments[i].fileName);
+      const nsname: string = resolveNsName(sourcePaths, emptyFiles[i].fsPath);
+      const snippets: string[] = [];
+      if (nsname) {
+          snippets.push(`(ns ${nsname}.${filename}\$1)`);
+      } else {
+          snippets.push(`(ns ${filename}\$1)`);
+      }
+      snippets.push("")
+      const textEditor = await vscode.window.showTextDocument(textDocuments[i]);
+      textEditor.insertSnippet(new vscode.SnippetString(snippets.join("\n")));
+  }
+}
```

Comment:
We can make this a bit more functional like this, I think.
```suggestion
export default async function handleNewCljFiles(e: vscode.FileCreateEvent) {
  const emptyFiles = e.files.filter(file => !isCljFile(file));
  let sourcePaths: string[] = [];
  const client = namespace.getSession();
  if (client) {
    const resp = await client.classpath();
    if (resp && resp.classpath) {
      sourcePaths = resp.classpath.filter(p => fs.existsSync(p) && fs.lstatSync(p).isDirectory());
    }
  }
  emptyFiles.forEach(async file => {
    const textDocument = await vscode.workspace.openTextDocument(file);
    if (textDocument.getText()) {
      return; // We only want to handle empty files
    }
    const filename: string = resolveFileName(textDocument.fileName);
    const nsname: string = resolveNsName(sourcePaths, file.fsPath);
    const snippets: string[] = [];
    if (nsname) {
      snippets.push(`(ns ${nsname}.${filename}\$1)`);
    } else {
      snippets.push(`(ns ${filename}\$1)`);
    }
    snippets.push("");
    const textEditor = await vscode.window.showTextDocument(textDocument);
    textEditor.insertSnippet(new vscode.SnippetString(snippets.join("\n")));
  });
}
```

Reply:
Thanks for the idea. I think I can come up with something more functionalish. But first I want to simplify the solution a bit.

Reply:
Agree, a bit more functionalish is to prefer. We should write some about that on our [Coding Style](https://github.com/BetterThanTomorrow/calva/wiki/Coding-Style) page.

Reply:
@bpringe, @PEZ The suggested functionalish version has one drawback: it doesn't short circuit. If there are no files to process, the array is empty, it still send a message to the server. I tried my best to make it more functionalish, but I failed, it's hard here because of presence of async code. If you write something like `cljFiles.filter(f =? isCljFile(f)).map(async fn)...`, you have to deal with array of promises, and also filter nils then.

But I managed to simplified the code a bit and going to push it in a few minutes.